src/azul/indexer.py

```diff
@@ -246,39 +246,26 @@ def _read_contributions(self, tallies: Tallies) -> List[Contribution]:
         return contributions
 
     def _aggregate(self, contributions: List[Contribution]) -> List[Aggregate]:
-        # Group contributions by entity ID and bundle UUID
-        contributions_by_bundle: Mapping[Tuple[EntityID, BundleUUID], List[Contribution]] = defaultdict(list)
+        # Group contributions by entity and bundle UUID
+        contributions_by_bundle: Mapping[Tuple[EntityReference, BundleUUID], List[Contribution]] = defaultdict(list)
         tallies = Counter()
         for contribution in contributions:
-            contributions_by_bundle[contribution.entity.entity_id, contribution.bundle_uuid].append(contribution)
+            contributions_by_bundle[contribution.entity, contribution.bundle_uuid].append(contribution)
             # Track the raw, unfiltered number of contributions per entity
-            tallies[contribution.entity.entity_id] += 1
-
-        # Among the contributions by a particular bundle to a particular entity, select the contribution from latest
-        # version of that bundle. If the latest non-deleted bundle. Group selected contributions by entity type and ID.
-        contributions_by_entity = defaultdict(list)
-        for bundle_contributions in contributions_by_bundle.values():
-            effective_contribution_by_version = {}
-            for c in bundle_contributions:
-                if c.bundle_deleted:
-                    effective_contribution_by_version[c.bundle_version] = c
-                else:
-                    try:
-                        effective_contribution = effective_contribution_by_version[c.bundle_version]
-                    except KeyError:
-                        effective_contribution_by_version[c.bundle_version] = c
-                    else:
-                        if not effective_contribution.bundle_deleted:
-                            effective_contribution_by_version[c.bundle_version] = c
-
-            undeleted_contributions = [c for c in effective_contribution_by_version.values() if not c.bundle_deleted]
-            try:
-                contribution = max(undeleted_contributions, key=attrgetter('bundle_version'))
-            except ValueError:
-                # max will throw ValueError for empty iterable
-                pass
-            else:
-                contributions_by_entity[contribution.entity].append(contribution)
+            tallies[contribution.entity] += 1
+
+        # For each entity and bundle, find the most recent contribution that is not a deletion
+        contributions_by_entity: Mapping[EntityReference, List[Contribution]] = defaultdict(list)
+        for (entity, bundle_uuid), contributions in contributions_by_bundle.items():
+            contributions = sorted(contributions, key=attrgetter('bundle_version', 'bundle_deleted'), reverse=True)
+            for bundle_version, group in groupby(contributions, key=attrgetter('bundle_version')):
```

Comment:
@hannes-ucsc this looks great! I wish I'd known about groupby. I'm a little envious that you wrote this code ;)